File: zenoh-grpc-client-sdk/zenoh-grpc-cpp/examples/pub.cpp
#include <iostream>

#include "common.hpp"

int main(int argc, char** argv) {
    auto session = zenoh_grpc::Session::connect(example_endpoint(argc, argv));

    zenoh_grpc::Session::PublisherOptions pub_options;
    pub_options.encoding = "text/plain";
    pub_options.priority = zenoh_grpc::Priority::Data;

    auto publisher = session.declare_publisher("demo/example/cpp", pub_options);
    std::cout << "publisher ready on " << example_endpoint(argc, argv) << std::endl;

    for (int i = 0; i < 500000; ++i) {
        std::string message = "hello from cpp " + std::to_string(i);
        zenoh_grpc::Publisher::PutOptions put_options;
        put_options.encoding = "text/plain";
        publisher.put(std::vector<std::uint8_t>(message.begin(), message.end()), put_options);
        std::cout << "published: " << message
                  << " dropped=" << publisher.send_dropped_count() << std::endl;
        example_sleep_ms(500);
    }

    return 0;
}

File: zenoh-grpc-client-sdk/zenoh-grpc-python/examples/pub.py
import zenoh_grpc
import time

# with zenoh_grpc.Session.connect("unix:///tmp/zenoh-grpc.sock") as session:
# with zenoh_grpc.Session.connect("tcp://127.0.0.1:7335") as session:
with zenoh_grpc.Session.connect() as session:
    pub = session.declare_publisher("demo/example/python", 
                                    priority=zenoh_grpc.Priority.DATA) # optional, default is Priority.DATA
    print("pub established.")

    for i in range(5000000000):
        msg = f"hello from python {i}"
        pub.put(msg.encode(), 
                encoding="text/plain") # the encoding is just a hint for the queryable to know how to parse the payload, it does not affect the actual payload sent to the queryable.
        # non-blocking, will drop if the internal queue is full

        print("\npublished to demo/example/python:", msg)
        print("publish queue dropped count:", pub.send_dropped_count()) # print the number of dropped messages in publish queue, which can be used to monitor the health of the connection between the publisher and the gRPC server. If the dropped count is increasing, it means the gRPC server can not keep up with the publishing speed, and you may want to reduce the publishing frequency.

        time.sleep(0.5) # gRPC server can only handle 10_000Hz for all clients with no dropped messages on my machine. So do not send messages too fast but send in batch and low frequency, or you will see the dropped count increasing. You can also adjust the sleep time to see how it affects the dropped count.
